### ISSUE TITLE
GCS session does not support requests transport adapters

### ISSUE THREAD

**Issue by matthewtberry** [dask/gcsfs#253]
I would like to use a custom [requests transport adapter](https://requests.readthedocs.io/en/master/user/advanced/#transport-adapters) on the session with GCS.

The session object is managed by `GCSFileSystem` class so a caller has no control over it.

**Comment by martindurant** [comment]
If both `requests.Session` and `AuthorizedSession` can take an adapter, it might make sense to provide this option and pass through the kwarg. Would you need to `mount` the adapter? In any case, I'd be happy to see this in a PR, if you have the time.

**Comment by matthewtberry** [comment]
Yes, `AuthorizedSession` inherits from `requests.Session` and an adapter needs to be mounted. I can see about putting together a PR.  Do you think providing adapters through `**kwargs` would be the best approach?

**Comment by martindurant** [comment]
In s3fs we have an optional `session_kwargs=None` that gets passed on

**Comment by matthewtberry** [comment]
I don't see how adapters can be mounted in s3fs.  `S3FileSystem` accepts `kwargs` that are passed on to the botocore session.

`S3FileSystem` allows a pre-created session to be passed in and used instead of creating a new one.  That could also be done in `GCSFileSystem` to allow mounting adapters, or the adapters could be passed separately and mounted in `GCSFileSystem`.

**Comment by martindurant** [comment]
Sorry, I wasn't clear: I mean that in s3fs, there is a mechanism for passing kwargs to wherever a thing gets instantiated (because a lot of hops are required to get things running); so I would suggest that GCSFileSystem acquire a session_kwargs argument, which is passed as a whole to Session (or the auth variant) and can include the adapter(s). The only extra part is, having to mount the endpoints, since I don't think you can do this with any arg to `Session`.

**Comment by martindurant** [comment]
Note: gcsfs now uses aiohttp instead of requests. I imagine it's still possible to do what you want, and will still involve adding extra optional *_kwargs to __init__, and passing them to the client.
